README, CONTRIBUTING & CODE_OF_CONDUCT

Robadob opened the pull request:
- [x] README
- [x] CONTRIBUTING
- [x] CODE-OF_CONDUCT
- [x] LICENCE (CC-BY-4.0)

Contributing & code-of-conduct were auto generated by contributing.md, though I made significant changes/extensions to `CONTRIBUTING.MD`.

In theory the guide is very light-touch, as we don't want people to feel that contributing is a chore with hundreds of vague style rules to follow.

Closes #54

JostMigenda commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,157 @@
+## I Want To Contribute
+
+> ### Legal Notice <!-- omit in toc -->
+> When contributing to this project, you must agree that you have authored 100% of the content, that you have the necessary rights to the content and that the content you contribute may be provided under the same licence as the existing content.
```
Is there an implied “no LLM slop” here?

Robadob replied:
This was just what contributing.md spat out, I left it in because we probably should cover ourselves if someone is clearly plagiarising something they shouldn't be.

I'm not personally bothered by LLMs being used to improve drafts etc, I'd be surprised if an LLM could produce a decent article in the same style at this stage.

Robadob replied:
> We encourage the use of LLMs and other tools for improving readability of your drafts, however if a significant proportion of the content you are submitting is produced by generative AI this should be noted in your pull request.

Have added this to the end of the legal statement, it's roughly inline with my local university policy.

JostMigenda commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,157 @@
+
+- Use a **clear and descriptive title** for the issue to identify the suggestion.
+- Provide a **step-by-step description of the suggested enhancement** in as many details as possible.
+- **Describe the current behavior** and **explain which behavior you expected to see instead** and why. At this point you can also tell which alternatives do not work for you.
```
This is focused on software projects and doesn’t fully apply to our knowledge base. @Robadob, based on the suggestions from the RSECon workshop that you have looked through so far, do you have any points on what makes a helpful contribution that you could add here?

JostMigenda commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,157 @@
+  - `subcategory`, *(Optimisations only)* Ways to sub-categorise the optimisation, e.g. the library it corresponds to.
+  - `tags`, *(Optimisations only)* An array of terms which may help this guide be found when searching (with the website's crude JS search).
+  - `style`, *(Profiler's only)* The profiling styles the profiler supports (e.g., Function-Level, Line-Level, Memory, Timeline, Hardware-Metrics, Benchmarking)
+  - `website`, *(Profiler's only)* URL of the profiler's main website.
```
```suggestion
- Setup the YAML header:
  - `level`, The priority of the guide in search results. Make a judgement call, more common issues should have a level closer to 1.
  - `published`, By default this should be set to `true`.
  - `authors`, The name of the author/s that have contributed to the guide.
  - `name`, A short and general name which describes the optimisation or the profiler's name.
  - `language`, An array of languages that the guide refers to. *Be careful to ensure this matches the capitalisation of other guides for the language!*
  - `subcategory`, *(Optimisations only)* Ways to sub-categorise the optimisation, e.g. the library it corresponds to.
  - `tags`, *(Optimisations only)* An array of terms which may help this guide be found when searching (with the website's crude JS search).
  - `style`, *(Profilers only)* The profiling styles the profiler supports (e.g., Function-Level, Line-Level, Memory, Timeline, Hardware-Metrics, Benchmarking)
  - `website`, *(Profilers only)* URL of the profiler's main website.
```
typos and phrasing

JostMigenda commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,157 @@
+
+You can use existing guides as templates—for example, [`_optimisations/python/numba.md`](https://github.com/sig-rpc/sig-rpc.github.io/blob/master/_optimisations/python/numba.md) or [`_profilers/python/cprofile.md`](https://github.com/sig-rpc/sig-rpc.github.io/blob/master/_profilers/python/cprofile.md).
+
+Once written, open a pull request against the `master` branch.
```
Are we still using `master`? `main` has been the default for 5+ years and, at least for me, is pretty deeply ingrained into my muscle memory by now.

Robadob commented on .github/CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,69 @@
+## Enforcement
+
+Instances of abusive, harassing, or otherwise unacceptable behaviour may be
+reported to the community leaders responsible for enforcement at <sig-rpc-managers@society-rse.org>.
```
> Out of interest: Who currently receives emails sent to that address?

All four officers (+ Lyndsey Ballantyne, she set up the mailing list so formally owns it as I don't think a google group can be owned by someone outside the socrse org)

> Also, does [the Society’s CoC committee](https://society-rse.org/about/governance/#:~:text=Code%20of%20Conduct%20committee) have a role to play for SIGs?

Ah good spot, the [draft ToR](https://github.com/sig-rpc/policies/issues/1) (SocRSE notified us approved, but never handed us back the ToR with their section completed) states we're bound by [SocRSE CoC](https://society-rse.org/about/policies/code-of-conduct-for-trustees/).

*Skim reading that again, I should really send them an annual update (probably just the AGM minutes) and chase ToR approval. I think alot wasn't setup properly, as they were still creating the process for SIGs when we setup.*

JostMigenda replied:
To be clear, the Trustee CoC only applies to elected officers, not to any contributor, right?

Robadob replied:
Lyndsey appears to have confirmed it only appllies to officers, so I redrafted (with AI support) a minimal contributor CoC, which is now linked from CONTRIBUTING and /pages/about.md Happy to bin that and instead direct them to event CoC if you think that's a better interpretation for casual contributions.

> You asked about interpretation of the codes of conduct and who they relate to.

> Anyone participating in Society of RSE communities and activities are to follow the events code of conduct, the Steering Group should additional follow the ‘trustee code of conduct’ – and on that note, now that we have many SIG’s I will also ask for that policy to be reviewed and reworded to make it clearer that it is not only for trustees.

https://github.com/sig-rpc/policies/pull/7

Robadob commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,157 @@
+#### Before Submitting an Enhancement
+
+- Make sure that you are using the latest version.
+- Read the [documentation](https://sig-rpc.github.io/) carefully and find out if the functionality is already covered, maybe by an individual configuration.
```
Probably worth getting rid of latest version too, as fair to assume people will be looking at live site rather than local copy.

Robadob commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,157 @@
+- Use a **clear and descriptive title** for the issue to identify the suggestion.
+- Provide a **step-by-step description of the suggested enhancement** in as many details as possible.
+- **Describe the current behavior** and **explain which behavior you expected to see instead** and why. At this point you can also tell which alternatives do not work for you.
+- You may want to **include screenshots or screen recordings** which help you demonstrate the steps or point out the part which the suggestion is related to. You can use [LICEcap](https://www.cockos.com/licecap/) to record GIFs on macOS and Windows, and the built-in [screen recorder in GNOME](https://help.gnome.org/users/gnome-help/stable/screen-shot-record.html.en) or [SimpleScreenRecorder](http ...
```
Wondering if i glossed over and missed reviewing this bit completely tbh, and got distracted writing the following section.

Robadob replied:
#### How Do I Make a Good Suggestion?

Suggested improvements are tracked as [GitHub issues](https://github.com/sig-rpc/sig-rpc.github.io/issues).

- Select and follow the **[most appropriate issue template](https://github.com/sig-rpc/sig-rpc.github.io/issues/new/choose)**.
- Use a **clear and descriptive title** for the issue to identify your suggestion.
- Be **specific about the location** on the website that your improvement corresponds to.
- **Explain why this improvement would be useful** to most Reasonable Performance Computing SIG's Website & Knowledgebase users.
- You may also want to point out **relevant websites/projects** that cover the suggested idea, which could serve as inspiration. 
- If you think a dynamic feature of the website is broken or could be improved:
  - **Describe the current behaviour** and **explain which behaviour you expected to see instead** and why.
  - **Include the web browser you are using**, and any other hardware/software details that may be relevant (e.g. is it caused by an ad-blocker?).

JostMigenda commented on .github/CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,69 @@
+# Code of Conduct - Reasonable Performance Computing SIG's Website & Knowledge-base
```
Maybe on the [About page](https://sig-rpc.github.io/about/)? That already has a link to the policies repo (and a—currently broken!—link to the ToR document), so it would fit in that paragraph. But beyond that, since the CoC is only applicable to people contributing/commenting/attending events, but not to anyone just visiting the website, I don’t think it needs to be very prominent on the home page.

Robadob commented on .github/CONTRIBUTING.md:
```diff
@@ -0,0 +1,157 @@
+- Make sure that you are using the latest version.
+- Read the [documentation](https://sig-rpc.github.io/) carefully and find out if the functionality is already covered, maybe by an individual configuration.
+- Perform a [search](https://github.com/sig-rpc/sig-rpc.github.io/issues) to see if the enhancement has already been suggested. If it has, add a comment to the existing issue instead of opening a new one.
+- Find out whether your idea fits with the scope and aims of the project. It's up to you to make a strong case to convince the project's developers of the merits of this feature. Keep in mind that we want features that will be useful to the majority of our users and not just a small subset. If you're just targeting a minority of users, consider writing an add-on/plugin library.
```
Fully reworded.

> - Check the [website](https://sig-rpc.github.io/), carefully to ensure your idea isn't redundant.
> - Perform a [search of existing issues](https://github.com/sig-rpc/sig-rpc.github.io/issues) to see if a similar suggestion has already been made. If it has, add a comment to the existing issue to show your support and add additional information, instead of opening a new one.
> - If you're suggesting a big change (more significant than a new guide), please ensure your idea fits with the scope and aims of SIG-RPC. It's up to you to make a strong case to convince the SIG's officers of the merits of your idea.

Robadob commented:
Hi @JostMigenda.

Thanks again for giving it such a review so quickly. I've now gone through the entire `CONTRIBUTING.MD` and have fixed all the issues I spotted along with those that you mentioned.

I've also moved it into the root of the repo as you suggested, and deleted the code of conduct, instead linking to socrse's, and added a short note about the CoC on the about page.

Likewise, as you should have seen via the officers email in cc, I've sent the (previously) relevant trustees an annual update and chased the final copy of the terms of reference so it can be properly installed into our org's policies repository.

Robadob commented on pages/about.md:
```diff
@@ -42,6 +42,11 @@ The current officers of SIG-RPC:
 
+## Code of Conduct
+
+As a SIG within Society RSE, SIG-RPCand it's members and contributors are bound [Society RSE's Organisation Code of Conduct][org_coc], with events additionally held under the [Society RSE Events Code of Conduct][event_coc]. Minor issues may be directed at the officers of SIG-RPC via [sig-rpc-managers@society-rse.org][officers_email] as a first point of call.
```
Have commit, will leave as unresolved until we have clarification from Lyndsey

ptheywood commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,168 @@
+
+### Your First Guide Contribution
+
+Guides live under the `optimisations/` and `profilers/` directories respectively, grouped by language. Each guide is written in Markdown (`.md`) and should:
```
Is it worth specifying that it's Jekyll/Kramdown markdown? (given there's a lot of variants, and the jekyll website is the target medium rather than viewing in github). 

I.e. `{:target="_blank"}` is a kramdown feature that won't behave consistently across other markdodn implemetnations?

Robadob replied:
```suggestion
Guides live under the `optimisations/` and `profilers/` directories respectively, grouped by language. Each guide is written in the [Kramdown](https://kramdown.gettalong.org/documentation.html) superset of Markdown (`.md`) and should:
```

ptheywood commented on README.md:
```diff
@@ -0,0 +1,53 @@
+The Reasonable Performance Computing SIG's Website & Knowledge-base, where user-submitted guides can help those writing software to identify and address performance problems in their code.
+
+The goal of the website is to make profiling and optimisation guidance accessible to even novice programmers. The main pre-requisite for readers is that they're able to read through code and recognise how it will execute.
+
```
An actual link to the website could be useful in the readme? 

Could be:
- standalone
- just adding a hyperlink to the h1
- or as part of the first sentence (`[Website & knowledge base](https://sig-rpc.github.io)`)?

Not neccessary, just a suggestion from viewing the readme file in isolation (rather than from the repo root)

ptheywood commented on README.md:
````diff
@@ -0,0 +1,53 @@
+
+```sh
+bundle install
+```
````
```suggestion
```sh
bundle config set path vendor/bundle
bundle install
```
```

> The location on disk where all gems in your bundle will be located regardless of $GEM_HOME or $GEM_PATH values. Bundle gems not found in this location will be installed by bundle install. Defaults to .bundle relative to repository root in Bundler 4, and to the default system path (Gem.dir) before Bundler 4.


I.e. for bundler < 4 this should prevent package conflicts between multiple jekyll websties on a given machine (assuming not also using rbenv or similar?)

Robadob commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,168 @@
+## I Want To Contribute
+
+> ### Legal Notice <!-- omit in toc -->
+> When contributing to this project, you must agree that you have authored 100% of the content, that you have the necessary rights to the content and that the content you contribute may be provided under the same licence as the existing content. We encourage the use of LLMs and other tools for improving readability of your drafts, however if a significant proportion of the content you are submitt ...
```
I would propose CC-BY, e.g. creative commons +  attribution.

I haven't set that up currently, because the codee catalogue suggested by someone from York where we may syndicate (tbc) elements is Apache licenced, so need to review compatibility.

ptheywood replied:
Yeah a CC licence seems most appropriate to me as well.

Apache is permissive (with some patent related stuff not inlcuded in MIT IIRC?) so should be fine, but would just need to make sure that those sections are marked as being apache, providing correct attribution etc.

Copyleft (GPL) snippets would need more care/thought, although the website wouldn't be linking against them...

Robadob commented on README.md:
```diff
@@ -0,0 +1,53 @@
+
+The Reasonable Performance Computing SIG's Website & Knowledge-base, where user-submitted guides can help those writing software to identify and address performance problems in their code.
+
+The goal of the website is to make profiling and optimisation guidance accessible to even novice programmers. The main pre-requisite for readers is that they're able to read through code and recognise how it will execute.
```
```suggestion
The goal of [the website](sig-rpc.github.io) is to make profiling and optimisation guidance accessible to even novice programmers. The main pre-requisite for readers is that they're able to read through code and recognise how it will execute.
```

Robadob commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,168 @@
+
+### Your First Guide Contribution
+
+Guides live under the `optimisations/` and `profilers/` directories respectively, grouped by language. Each guide is written in Markdown (`.md`) and should:
```
Outdated, but done seperatley

JostMigenda commented:
> Hi Jost, I _think_ everything is addressed now.
> 
> See remaining open threads for last remaining decision regarding CoC.

Many thanks! Yes, I agree everything else is done now. Following discussion in https://github.com/sig-rpc/policies/pull/7, I’ve added three suggestions for updating the about and CONTRIBUTING pages accordingly. If there’s no new discussion around those, then I think this PR will be ready to merge afterwards.

> The two PRs in https://github.com/sig-rpc/policies/pulls?q=sort%3Aupdated-desc+is%3Apr+is%3Aopen need to be merged prior to all the links in here being valid.

Those are ready, too. 👍

> One approved I'll clean up commit history and notify all prior contributors via comments in pull requests.

As part of the clean up, can you also drop [the MATLAB commit](https://github.com/sig-rpc/sig-rpc.github.io/pull/78/commits/d23368fc76a54fe3c40798242f1f08d360c87b4b) that somehow slipped in? That was already merged in #80, so looks like a git mishap.
